### ISSUE TITLE
Semantic highlighting in source

### ISSUE THREAD

**Issue by bruno-medeiros** [GoClipse/goclipse#111]
Highlight source code according to semantic properties, example:

-function declaration names
-type declaration names
-variable names

**Comment by rillig** [comment]
+1

**Comment by 10liuguang01** [comment]
Come on

**Comment by rsippl** [comment]
+1

This lacking feature is the only reason that keeps me from switching from IntelliJ Idea.

**Comment by bruno-medeiros** [comment]
How does the semantic highlighting work in IntelliJ, which semantic elements are highlighted?

**Comment by rsippl** [comment]
Here's what it looks like:

![idea-go-1](https://cloud.githubusercontent.com/assets/260389/15244804/43dd1b38-1903-11e6-988f-65dbfbaabd3d.png)
![idea-go-2](https://cloud.githubusercontent.com/assets/260389/15244807/49687228-1903-11e6-9454-205011c967ca.png)
![idea-go-3](https://cloud.githubusercontent.com/assets/260389/15244811/4dd691e6-1903-11e6-93fe-9188ab169d07.png)

**Comment by rsippl** [comment]
[Here's](https://github.com/go-lang-plugin-org/go-lang-idea-plugin) the plugin code (not sure if it helps)

**Comment by bruno-medeiros** [comment]
Yeah, I know about go-lang-idea-plugin already. I'm not sure I quite understand the logic of that highlighting though. `Println` is a function and is highlighted red, same as `RGBA` which is a struct. At the same time, the methods `ScaleBy` and  `Distance`  are not highlighted...

**Comment by sanmai** [comment]
Here's how it looks like in Eclipse's PDE:

![highlight](https://cloud.githubusercontent.com/assets/139488/16575796/e2a21322-42c5-11e6-9b07-e4493d570729.png)
(note the cursor position)

**Comment by rillig** [comment]
IntelliJ knows the following token categories for coloring Go programs:

* Bad character
* Block comment
* Braces
* Brackets
* Builtin function
* Builtin type
* Colon
* Comma
* Dot
* Exported function
* Function parameter
* Identifier
* Keyword
* Label
* Line comment
* Local {constant, function, variable}
* Method receiver
* Number
* Operator
* Package exported {constant, interface, struct, variable}
* Package local {constant, interface, struct, variable}
* Parentheses
* Scope declared variable
* Semicolon
* String
* Struct exported member
* Struct local member
* Type reference
* Type specification

Some additional categories I find useful are:

* Local variable declaration
* Local variable use
* Local variable overwrite
* Control flow keyword (if, for, switch, return, goto)
* Anonymous variable (_)
* Imported package (e.g. `fmt`)

When you install IntelliJ, you can find many more token categories in the Settings dialog (Alt+Ctrl+S), section Editor > Colors & Fonts. Note the General section, which has entries for “Unused symbol” and similar things.
